common-components/proto/api/type/decimal.proto

```diff
@@ -0,0 +1,36 @@
+syntax = "proto3";
+
+package api.type;
+
+option cc_enable_arenas = true;
+option go_package = "aip.dev/type/decimal;decimal";
+option java_multiple_files = true;
+option java_outer_classname = "DecimalProto";
+option java_package = "dev.aip.type";
+option objc_class_prefix = "AIPTYPE";
+
+// Represents a decimal number in a form similar to scientific notation.
+//
+// Examples:
+// 17            === {digits: 17   exponent: 0} (or just {digits: 17})
+// -0.005        === {digits: -5   exponent: -3}
+// 33.5 million  === {digits: 335  exponent: 5}
+// 11/8 (1.375)  === {digits: 1375 exponent: -3}
+message Decimal {
+  // The significant digits (a.k.a. significand or mantissa) of the number.
+  int64 digits = 1;
+
+  // Represents the position of the decimal point within the digits.
+  //
+  // When the exponent is 0, the value of the Decimal is simply the value of
+  // `digits`.
+  //
+  // When the exponent is > 0, represents the number of trailing zeroes after
+  // the digits.
+  //
+  // When the exponent is < 0, represents how many of the significant digits
+  // (and implicit leading zeroes, as needed) come after the decmial point.
+  //
+  // When unset, this value is implicitly zero (0).
+  int64 exponent = 2;
```

Comment:
This isn't really a "pure" exponent. It's a power of 10. I think "e" is the least bad term (with an appropriate comment). The rust_decimal crate uses "scale".

Reply:
From https://en.wikipedia.org/wiki/Scientific_notation:

> In scientific notation, nonzero numbers are written in the form
>
> *m × 10n*
>
> or m times ten raised to the power of n, where n is an integer, and the coefficient m is a nonzero real number (usually between 1 and 10 in absolute value, and nearly always written as a terminating decimal). The integer n is called the exponent...


(This is more relevant to the question of "digits" vs. "mantissa", but in the interest of completeness and not seeming like I'm cutting off the quote at a cherrypicked point, that section goes on to talk about usage of "significand and mantissa" for *m*, and then goes on to say that "mantissa" is ambiguous.)

Reply:
Also, it looks like in both Rust and Java, `scale` is the inverse of exponent (i.e. an exponent of 2 is equivalent to a scale of -2).  Do I have that right?

Reply:
> but in the interest of completeness and not seeming like I'm cutting off the quote at a cherrypicked point

Nah, you're fine, and this is persuasive.

> Also, it looks like in both Rust and Java, scale is the inverse of exponent (i.e. an exponent of 2 is equivalent to a scale of -2). Do I have that right?

Yes.

Reply:
I find this persuasive.